File: scripts/review-prompt.js
#!/usr/bin/env node
const { execFileSync } = require('child_process');
const fs = require('fs');
const path = require('path');
const {
  ensureRuntime,
  getProjectRoot,
  getUnreviewedFiles,
  loadSettings,
  loadState,
  markFilesReviewed
} = require('./state');

function shellGitDiff(files, projectRoot) {
  try {
    return execFileSync('git', ['diff', '--', ...files], {
      cwd: projectRoot,
      encoding: 'utf8',
      maxBuffer: 10 * 1024 * 1024
    });
  } catch (error) {
    const stderr = error && error.stderr ? String(error.stderr).trim() : '';
    return `Git diff unavailable. Review the current file contents directly.\n${stderr}`;
  }
}

function readIfExists(filePath, fallback = '') {
  return fs.existsSync(filePath) ? fs.readFileSync(filePath, 'utf8') : fallback;
}

function currentFileSnapshots(files, projectRoot) {
  return files.map((file) => {
    const fullPath = path.join(projectRoot, file);
    if (!fs.existsSync(fullPath) || !fs.statSync(fullPath).isFile()) {
      return `## ${file}\n\nFile does not currently exist.`;
    }
    const content = fs.readFileSync(fullPath, 'utf8');
    const maxChars = 40000;
    const truncated = content.length > maxChars
      ? `${content.slice(0, maxChars)}\n\n[truncated at ${maxChars} characters]`
      : content;
    return `## ${file}\n\n\`\`\`\n${truncated}\n\`\`\``;
  }).join('\n\n');
}

function writeProjectScriptShim(projectRoot, pluginScriptPath) {
  const runtimeScripts = path.join(projectRoot, '.claude', 'claude-auto-review', 'scripts');
  fs.mkdirSync(runtimeScripts, { recursive: true });
  const shimPath = path.join(runtimeScripts, 'review-prompt.js');
  const relative = path.relative(runtimeScripts, pluginScriptPath).split(path.sep).join('/');
  const content = `#!/usr/bin/env node\nrequire(${JSON.stringify(relative)});\n`;
  if (!fs.existsSync(shimPath) || fs.readFileSync(shimPath, 'utf8') !== content) {
    fs.writeFileSync(shimPath, content);
  }
}

function buildPrompt({ reviewId, timestamp, files, rules, diff, snapshots, reviewPath }) {
  const fileList = files.map((entry) => `- ${entry.file} (hash: ${entry.hash})`).join('\n');
  return `# Claude Auto Review Request ${reviewId}

You must review the changed files before stopping. Use the reviewer agent behavior from \`agents/reviewer.md\`: focus on semantic bugs, security, maintainability, and project rules. Do not nitpick formatting.

## Review Output

Write the final review to:

\`${reviewPath}\`

Use this exact top matter:

\`\`\`markdown
# Review ${reviewId} - ${timestamp}

## Files Reviewed
${fileList}

## Findings
\`\`\`

If no findings exist, write "Clean - no issues found. Claude may stop." under "## Verdict".

## Files To Review
${fileList}

## Rules
${rules}

## Git Diff
\`\`\`diff
${diff}
\`\`\`

## Current File Snapshots
${snapshots}

## After Review

Fix any CRITICAL or HIGH findings you agree with. If you edit files, the hook will track those new hashes and require another review pass.`;
}

function main() {
  try {
    const projectRoot = getProjectRoot();
    const runtime = ensureRuntime(projectRoot);
    writeProjectScriptShim(projectRoot, __filename);

    const settings = loadSettings(projectRoot);
    if (!settings.enabled) {
      console.log('Claude Auto Review is disabled in .claude/settings.json.');
      process.exit(0);
    }

    const unreviewed = getUnreviewedFiles(loadState(projectRoot));
    if (unreviewed.length === 0) {
      console.log('Claude Auto Review: no unreviewed changes.');
      process.exit(0);
    }

    const timestamp = new Date().toISOString();
    const reviewId = `rev-${timestamp.replace(/[-:.TZ]/g, '').slice(0, 14)}`;
    const files = unreviewed.map((entry) => entry.file);
    const configuredRules = settings.rulesFile || runtime.rulesPath;
    const rulesPath = path.isAbsolute(configuredRules) ? configuredRules : path.join(projectRoot, configuredRules);
    const rules = readIfExists(rulesPath, readIfExists(runtime.rulesPath));
    const diff = shellGitDiff(files, projectRoot);
    const snapshots = currentFileSnapshots(files, projectRoot);

    const reviewPath = path.join(runtime.baseDir, 'reviews', `review-${reviewId}.md`);
    const promptPath = path.join(runtime.baseDir, 'run', `review-${reviewId}-prompt.md`);
    const prompt = buildPrompt({
      reviewId,
      timestamp,
      files: unreviewed,
      rules,
      diff,
      snapshots,
      reviewPath
    });

    fs.writeFileSync(promptPath, prompt);
    fs.writeFileSync(reviewPath, `# Review ${reviewId} - ${timestamp}

## Files Reviewed
${unreviewed.map((entry) => `- ${entry.file} (hash: ${entry.hash})`).join('\n')}

## Findings

Pending. Claude must complete this review from ${promptPath}.

## Verdict

Pending.
`);

    markFilesReviewed(unreviewed, reviewId, projectRoot);

    console.log(`Claude Auto Review prompt created: ${promptPath}`);
    console.log(`Review file initialized: ${reviewPath}`);
    console.log('Read the prompt, complete the review file, and fix any agreed CRITICAL or HIGH findings before stopping.');
  } catch (error) {
    console.error(`Claude Auto Review failed open: ${error.message}`);
    process.exit(0);
  }
}

main();

File: scripts/review_prompt.py
#!/usr/bin/env python3
from pathlib import Path

from bootstrap import ensure_repo_root_on_path

ensure_repo_root_on_path()

from scripts.paths import get_client_id, get_project_root
from scripts.review_prompt_flow import create_review_prompt_files
from scripts.shims import write_project_script_shim
from scripts.state import (
    append_review_started,
    ensure_client_runtime,
    get_unreviewed_files,
    log_event,
    load_settings,
    load_state,
)


def main():
    try:
        project_root = get_project_root()
        client_id = get_client_id()
        ensure_client_runtime(project_root, client_id)
        write_project_script_shim(
            Path(project_root) / ".claude" / "claude-auto-review" / "scripts" / "review_prompt.py",
            Path(__file__).resolve(),
        )

        settings = load_settings(project_root)
        if not settings.get("enabled", True):
            log_event(project_root, "review_prompt_disabled")
            print("Claude Auto Review is disabled in .claude/settings.json.")
            return 0

        unreviewed = get_unreviewed_files(load_state(project_root, client_id))
        if not unreviewed:
            log_event(project_root, "review_prompt_noop")
            print("Claude Auto Review: no unreviewed changes.")
            return 0

        artifacts = create_review_prompt_files(project_root, client_id, unreviewed, settings)

        append_review_started(unreviewed, artifacts.review_id, artifacts.review_path, project_root, client_id=client_id)
        log_event(
            project_root,
            "review_prompt_created",
            reviewId=artifacts.review_id,
            files=artifacts.files,
            prompt=str(artifacts.prompt_path),
            review=str(artifacts.review_path),
            clientId=client_id,
        )
        print(f"Claude Auto Review prompt created: {artifacts.prompt_path}")
        print(f"Review file initialized: {artifacts.review_path}")
        print("Read the prompt, complete the review file, and fix any agreed CRITICAL or HIGH findings before stopping.")
        return 0
    except Exception as error:
        try:
            log_event(get_project_root(), "review_prompt_error", error=str(error))
        except Exception:
            pass
        print(f"Claude Auto Review failed open: {error}")
        return 0


if __name__ == "__main__":
    raise SystemExit(main())
